client.go

```diff
@@ -250,74 +249,65 @@ func (np *PubNameAndPassword) PubKey() (e4crypto.Ed25519PublicKey, error) {
 // depending the given ClientConfig
 //
 // config is a ClientConfig, either SymIDAndKey, SymNameAndPassword, PubIDAndKey or PubNameAndPassword
-// persistStatePath is the file system path to the file to read and persist the client's state.
-func NewClient(config ClientConfig, persistStatePath string) (Client, error) {
-	return config.genNewClient(persistStatePath)
+// store is an e4.Store implementation
+func NewClient(config ClientConfig, store Store) (Client, error) {
+	return config.genNewClient(store)
 }
 
 // newClient creates a new client, generating a random ID if they are empty
-func newClient(id []byte, clientKey keys.KeyMaterial, persistStatePath string) (Client, error) {
+func newClient(id []byte, clientKey keys.KeyMaterial, store Store) (Client, error) {
 	if len(id) == 0 {
 		return nil, errors.New("client id must not be empty")
 	}
 
 	c := &client{
 		Key:            clientKey,
 		TopicKeys:      make(map[string]keys.TopicKey),
-		FilePath:       persistStatePath,
 		ReceivingTopic: TopicForID(id),
+
+		store: store,
 	}
 
 	c.ID = make([]byte, len(id))
 	copy(c.ID, id)
 
 	log.SetPrefix("e4client\t")
 
+	if err := c.save(); err != nil {
+		return nil, err
+	}
+
 	return c, nil
 }
 
 // LoadClient loads a client state from the file system
-func LoadClient(persistStatePath string) (Client, error) {
+func LoadClient(store Store) (Client, error) {
 	c := &client{}
-	err := readJSON(persistStatePath, c)
+
+	decoder := json.NewDecoder(store)
+	err := decoder.Decode(c)
 	if err != nil {
 		return nil, err
 	}
 
+	store.Seek(0, io.SeekStart)
+
+	c.store = store
+
 	return c, nil
 }
 
 func (c *client) save() error {
-	err := writeJSON(c.FilePath, c)
+	encoder := json.NewEncoder(c.store)
```

Comment:
So if I understand correctly, with the memory store this will serialize `Client` to json and dump it in a memory blob.

Do we want to necessarily always use json for backing storage? I honestly don't know the answer to that, and I think the answer is yes for files and for html5 localStorage (for e4go at least). The libe4 "mem" implementation just does a no-op on save and load (storages in libe4 are responsible for their own load/save logic, if this makes sense in that use case). Of course mem is just a toy implementation so... I dunno. What do you think?

Reply:
I think the serialization must stay inside the client. Delegating it to the store would open lots of potential  issues and complexity in the store implementation,  by having to deal with the full client struct rather than just raw []byte. 

Now for it being json, this is a good question. I don't really know, but we could allow users to set their own encoder in place of the json one (that we can keep as a default), in case json doesn't fit for any reasons.